### DISCUSSION TITLE
`hooks.read-source-state` discussion

### DISCUSSION THREAD

**Discussion by twpayne** [twpayne/chezmoi#3342] [Show and tell]
#3343 adds a `read-source-state` hook that allows you to run an arbitrary command before chezmoi reads the source state. This can be used, for example, to install your password manager during `chezmoi init` on a new machine.

This is a powerful feature and is currently in beta.

If you use this hook, please describe any problems you encounter, or how it can be improved here.

**Comment by arrrgi** [comment]
Adding a comment to this discussion as the intended effects of the new hook may or may not be too wide in scope.

Having tested this now with https://github.com/arrrgi/dotfiles/blob/719c26d4ba0c10636d7d091e2e27338f4785f746/home/.chezmoi.yaml.tmpl#L159-L162 and https://github.com/arrrgi/dotfiles/blob/feature/setup-scripts/home/.hooks/.install-hcp-vlt.sh

I could see that the read-source-state hook was also executed when running `chezmoi data`. I'm not sure if this is an expected/wanted/unwanted side effect but it was a surprise nonetheless.

For me the impact was nil as my script is idempotent(_ish_) but it may catch others out.

**Comment by Lockszmith-GH** [comment]
Very interesting!

~~How would I deal with multi-platform? let's say `powershell.exe` on windows, and `sh` on linux/unix?~~

Sorry, it's still a template, dumb question. Will provide feedback when I get to test it.

**Comment by twpayne** [comment]
Thanks for the feedback!

> I could see that the read-source-state hook was also executed when running `chezmoi data`. I'm not sure if this is an expected/wanted/unwanted side effect but it was a surprise nonetheless.

This is expected, as the source state can contain `.chezmoidata` files, so the source state has to be read by the `chezmoi data` command.

**Comment by dgudim** [comment]
How do I get the directory chezmoi was run in from a hook?

**Reply by twpayne** [reply]
chezmoi always runs hooks with your home directory as the working directory. Why do you need to know which directory chezmoi was run in?

**Reply by dgudim** [reply]
I want to do some checks before adding the file to chezmoi and I can't see where the file is if I run `chezmoi add` in the same dir as the file

**Reply by twpayne** [reply]
OK. #3390 adds a `CHEZMOI_COMMAND_DIR` environment variable containing the directory that chezmoi was run in. Note that for `chezmoi add` you should use `hooks.add.pre`.
